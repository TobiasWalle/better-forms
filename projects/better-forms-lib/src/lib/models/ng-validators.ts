import { AsyncValidator, AsyncValidatorFn, Validator, ValidatorFn } from '@angular/forms';

export type SyncNgValidator = Validator | ValidatorFn;
export type AsyncNgValidator = AsyncValidator | AsyncValidatorFn;
export type NgValidator = SyncNgValidator | AsyncValidatorFn;
