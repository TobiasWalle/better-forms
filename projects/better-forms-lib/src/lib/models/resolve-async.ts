import { isObservable, Observable } from 'rxjs';

export async function resolveAsync<T>(value: T | Promise<T> | Observable<T>): Promise<T> {
  if (isObservable(value)) {
    return await (value as Observable<T>).toPromise();
  } else {
    return value;
  }
}
