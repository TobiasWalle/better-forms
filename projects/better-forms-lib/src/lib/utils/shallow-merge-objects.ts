export function shallowMergeObjects<T>(objects: T[]): T {
  return objects.reduce((object, nextObject) => {
    return {
      ...object,
      ...nextObject
    };
  }, {});
}
