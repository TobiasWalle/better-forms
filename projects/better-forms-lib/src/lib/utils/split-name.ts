export function splitName(path: string): string[] {
  return path.split('.');
}
