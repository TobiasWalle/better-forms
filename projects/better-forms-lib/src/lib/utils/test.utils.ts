import { ControlValueAccessor } from '@angular/forms';

export function createMockValueAccessor(): ControlValueAccessor {
  return {
    writeValue: jest.fn(),
    registerOnChange: jest.fn(),
    registerOnTouched: jest.fn(),
    setDisabledState: jest.fn(),
  };
}
