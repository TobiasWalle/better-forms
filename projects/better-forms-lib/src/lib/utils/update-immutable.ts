const immutable: any = require('object-path-immutable');

export function setImmutable<T>(original: T, path: string[], newValue: any): T {
  return immutable.set(original, path, newValue);
}

export function getByPath<T>(object: T, path: string[]): any {
  return path.reduce((nextObject: any, key: string) => {
    if (key in nextObject) {
      return nextObject[key];
    }
    throw new Error(`Couldn't find "${path}" on object. "${key}" doesn not exists on ${nextObject}.`);
  }, object);
}

