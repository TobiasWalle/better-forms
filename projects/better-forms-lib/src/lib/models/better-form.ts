import { BehaviorSubject, Observable, Subject } from 'rxjs';
import { getByPath, setImmutable } from '../utils/update-immutable';

interface UpdateInformation {
  path: string[];
  newValue: any;
}

interface BetterFormOptions<T> {
  initialValue: T;
}

export class BetterForm<T> {
  private valueSubject: BehaviorSubject<T>;
  public valueChange: Observable<T>;

  private onUpdateSubject: Subject<UpdateInformation> = new Subject<UpdateInformation>();
  public onUpdate: Observable<UpdateInformation> = this.onUpdateSubject.asObservable();

  public get value(): T {
    return this.valueSubject.value;
  }

  constructor(options: BetterFormOptions<T>) {
    this.valueSubject = new BehaviorSubject<T>(options.initialValue);
    this.valueChange = this.valueSubject.asObservable();
  }

  public setValue(value: T): void {
    this.triggerUpdate([], value);
  }

  public updatePath(path: string[], value: any): void {
    this.triggerUpdate(path, value);
  }

  public getByPath(path: string[]): any {
    return getByPath(this.value, path);
  }

  private triggerUpdate(path: string[], newValue: any): void {
    const newFormValue = setImmutable(this.value, path, newValue);
    this.valueSubject.next(newFormValue);
    this.onUpdateSubject.next({ path, newValue });
  }
}
