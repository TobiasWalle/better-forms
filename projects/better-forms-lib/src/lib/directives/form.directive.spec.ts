import { BetterForm } from '../models/better-form';
import { createMockValueAccessor } from '../utils/test.utils';
import { FormDirective } from './form.directive';

interface User {
  name: {
    first: string;
    last: string;
  };
  age: number;
}

describe('FormDirective', () => {
  const initialValue: User = {
    name: {
      first: 'Michael',
      last: 'Schmidt'
    },
    age: 30
  };
  let directive: FormDirective;

  beforeEach(() => {
    directive = new FormDirective();
    directive.betterForm = new BetterForm<User>({ initialValue });
  });

  it('should create an instance', () => {
    expect(directive).toBeTruthy();
  });

  it('should throw an error if the betterForm input is not set', () => {
    directive.betterForm = undefined;

    expect(() => directive.ngOnInit()).toThrowErrorMatchingSnapshot();
  });

  it('should register value accessors, update values and unregister', () => {
    directive.ngOnInit();

    const firstName = createMockValueAccessor();
    const age = createMockValueAccessor();
    directive.registerControlValueAccessor('name.first', firstName);
    directive.registerControlValueAccessor('age', age);

    expect(firstName.writeValue).toHaveBeenCalledWith('Michael');
    expect(age.writeValue).toHaveBeenCalledWith(30);
    expect(firstName.writeValue).toHaveBeenCalledTimes(1);
    expect(age.writeValue).toHaveBeenCalledTimes(1);

    directive.betterForm.updatePath(['age'], 21);

    expect(age.writeValue).toHaveBeenCalledWith(21);
    expect(firstName.writeValue).toHaveBeenCalledTimes(1);
    expect(age.writeValue).toHaveBeenCalledTimes(2);

    directive.betterForm.updatePath(['name', 'first'], 'Test');
    directive.betterForm.updatePath(['name', 'last'], 'LastName');

    expect(firstName.writeValue).toHaveBeenCalledWith('Test');
    expect(firstName.writeValue).toHaveBeenCalledTimes(2);
    expect(age.writeValue).toHaveBeenCalledTimes(2);

    directive.unregisterControlValueAccessor('age');

    directive.betterForm.updatePath(['age'], 20);
    expect(age.writeValue).toHaveBeenCalledTimes(2);
  });

  it('should throw error if valueAccessor is registered twice', () => {
    directive.ngOnInit();

    const age = createMockValueAccessor();
    directive.registerControlValueAccessor('age', age);

    expect(() => directive.registerControlValueAccessor('age', age)).toThrowErrorMatchingSnapshot();
  });
});
