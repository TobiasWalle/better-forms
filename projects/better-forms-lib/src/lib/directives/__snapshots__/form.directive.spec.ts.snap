// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormDirective should throw an error if the betterForm input is not set 1`] = `"The Input \\"betterForm\\" is required"`;

exports[`FormDirective should throw error if valueAccessor is registered twice 1`] = `"Value Accessor with name \\"age\\" already registered"`;
